File: CS/WindowsFormsApplication202/ImageSliderHelper.cs
using DevExpress.Utils;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.CustomEditor;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Grid;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;

namespace WindowsFormsApplication202 {
    static public class ImageSliderHelper {
        static GridView view = null;
        static ImageCollection imageList = null;
        static bool canSaveFirstImage = false;
        static bool isFirstImageFound = false;
        static Image firstImage = null;
        static int currentIndex = -1;

        public static void BindImageSlider(GridControl grid, string columnFieldName, ImageCollection images) {
            imageList = images;
            RepositoryItemAnyControl repository = new RepositoryItemAnyControl();
            grid.RepositoryItems.Add(repository);
            MyImageSlider slider = new MyImageSlider();
            view = grid.MainView as GridView;
            repository.Control = slider;
            view.Columns["Image"].ColumnEdit = repository;
            view.Columns["Description"].OptionsColumn.AllowEdit = false;
            view.RowHeight = slider.Size.Height;
            view.ShownEditor += view_ShownEditor;
        }

        private static void view_ShownEditor(object sender, EventArgs e) {
            GridView view = sender as GridView;
            AnyControlEdit edit = view.ActiveEditor as AnyControlEdit;
            if (edit != null) {
                MyImageSlider slider = edit.Properties.Control as MyImageSlider;
                slider.ImageList = imageList;
                OnEnter(slider);
                slider.MouseClick += slider_MouseClick;
                slider.Leave += slider_Leave;
            }
        }

        static void slider_Leave(object sender, EventArgs e) {
            ImageSlider slider = sender as ImageSlider;
            if (slider.CurrentImageIndex != -1 && slider.ImageList != null && canSaveFirstImage) {
                if (!isFirstImageFound) {
                    ((ImageCollection)slider.ImageList).Images.RemoveAt(0);
                }
                view.SetRowCellValue(view.FocusedRowHandle, view.FocusedColumn, (slider.ImageList as ImageCollection).Images[slider.CurrentImageIndex]);
            }
        }

        static void slider_MouseClick(object sender, System.Windows.Forms.MouseEventArgs e) {
            ImageSlider slider = sender as ImageSlider;
            ImageSliderHitInfo info = slider.CalcHitInfo(e.Location);
            if ((info.HitTest == ImageSliderHitTest.LeftButton || info.HitTest == ImageSliderHitTest.RightButton))
                canSaveFirstImage = true;
        }

        static void OnEnter(MyImageSlider slider) {
            firstImage = slider.Images[0];
            isFirstImageFound = false;
            if (imageList != null) {
                for (int i = 0; i < ((ImageCollection)imageList).Images.Count; i++) {
                    if (firstImage == ((ImageCollection)imageList).Images[i]) {
                        currentIndex = i;
                        isFirstImageFound = true;
                    }
                }
                if (currentIndex == -1) {
                    ((ImageCollection)imageList).Images.Insert(0, firstImage);
                    slider.CurrentImageIndex = 0;
                }
                if (currentIndex != -1) {
                    slider.CurrentImageIndex = currentIndex;
                }
            }
        }
    }
}

File: VB/WindowsFormsApplication202/ImageSliderHelper.vb
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.CustomEditor
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Grid
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Text

Namespace WindowsFormsApplication202
	Public NotInheritable Class ImageSliderHelper

		Private Sub New()
		End Sub

		Private Shared view As GridView = Nothing
		Private Shared imageList As ImageCollection = Nothing
		Private Shared canSaveFirstImage As Boolean = False
		Private Shared isFirstImageFound As Boolean = False
		Private Shared firstImage As Image = Nothing
		Private Shared currentIndex As Integer = -1

		Public Shared Sub BindImageSlider(ByVal grid As GridControl, ByVal columnFieldName As String, ByVal images As ImageCollection)
			imageList = images
			Dim repository As New RepositoryItemAnyControl()
			grid.RepositoryItems.Add(repository)
			Dim slider As New MyImageSlider()
			view = TryCast(grid.MainView, GridView)
			repository.Control = slider
			view.Columns("Image").ColumnEdit = repository
			view.Columns("Description").OptionsColumn.AllowEdit = False
			view.RowHeight = slider.Size.Height
			AddHandler view.ShownEditor, AddressOf view_ShownEditor
		End Sub

		Private Shared Sub view_ShownEditor(ByVal sender As Object, ByVal e As EventArgs)
			Dim view As GridView = TryCast(sender, GridView)
			Dim edit As AnyControlEdit = TryCast(view.ActiveEditor, AnyControlEdit)
			If edit IsNot Nothing Then
				Dim slider As MyImageSlider = TryCast(edit.Properties.Control, MyImageSlider)
				slider.ImageList = imageList
				OnEnter(slider)
				AddHandler slider.MouseClick, AddressOf slider_MouseClick
				AddHandler slider.Leave, AddressOf slider_Leave
			End If
		End Sub

		Private Shared Sub slider_Leave(ByVal sender As Object, ByVal e As EventArgs)
			Dim slider As ImageSlider = TryCast(sender, ImageSlider)
			If slider.CurrentImageIndex <> -1 AndAlso slider.ImageList IsNot Nothing AndAlso canSaveFirstImage Then
				If Not isFirstImageFound Then
					DirectCast(slider.ImageList, ImageCollection).Images.RemoveAt(0)
				End If
				view.SetRowCellValue(view.FocusedRowHandle, view.FocusedColumn, (TryCast(slider.ImageList, ImageCollection)).Images(slider.CurrentImageIndex))
			End If
		End Sub

		Private Shared Sub slider_MouseClick(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
			Dim slider As ImageSlider = TryCast(sender, ImageSlider)
			Dim info As ImageSliderHitInfo = slider.CalcHitInfo(e.Location)
			If (info.HitTest = ImageSliderHitTest.LeftButton OrElse info.HitTest = ImageSliderHitTest.RightButton) Then
				canSaveFirstImage = True
			End If
		End Sub

		Private Shared Sub OnEnter(ByVal slider As MyImageSlider)
			firstImage = slider.Images(0)
			isFirstImageFound = False
			If imageList IsNot Nothing Then
				For i As Integer = 0 To (CType(imageList, ImageCollection).Images.Count) - 1
					If firstImage Is CType(imageList, ImageCollection).Images(i) Then
						currentIndex = i
						isFirstImageFound = True
					End If
				Next i
				If currentIndex = -1 Then
					CType(imageList, ImageCollection).Images.Insert(0, firstImage)
					slider.CurrentImageIndex = 0
				End If
				If currentIndex <> -1 Then
					slider.CurrentImageIndex = currentIndex
				End If
			End If
		End Sub
	End Class
End Namespace
